### ISSUE TITLE
`callback_metrics` should be reset when a new epoch start

### ISSUE THREAD

**Issue by rohitgr7** [PyTorchLightning/pytorch-lightning#7558]
## 🐛 Bug

Callback metrics should be destroyed whenever a new epoch/loop starts.

### To Reproduce

```python
def test_bug(tmpdir):
    import pytorch_lightning as pl
    from tests.helpers.boring_model import BoringModel

    class TestModel(BoringModel):

        def on_train_start(self) -> None:
            assert not self.trainer.callback_metrics

        def training_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('train_loss', loss)
            return loss

        def validation_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('val_loss', loss)

    model = TestModel()
    model.validation_epoch_end = None
    trainer = pl.Trainer(
        default_root_dir=tmpdir,
        max_epochs=3,
        num_sanity_val_steps=0,
        callbacks=[pl.callbacks.ModelCheckpoint(monitor='val_loss')]
    )

    trainer.validate(model)
    print(trainer.callback_metrics)

    model.validation_step = None
    trainer.fit(model)
```

### Expected behavior

The above case should throw an error because during `.fit` validation is disabled but since `trainer.validate` runs before it was disabled, the callback_metrics calculated during that loop stays and doesn't cause a `MisconfigurationException` during checkpointing. This is just one case where wrong metrics stored can cause some issues.

Another one can occur if `check_val_every_n_epoch>1` where if validation is not happening in every epoch, it still contains metrics from some old epoch which can cause some issues with `ModelCheckpoint/EarlyStopping/ReduceLROnPlateau`.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1+cu101
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: Tesla T4

**Comment by rohitgr7** [comment]
Also noticed one more thing with checkpointing here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/e126649d19e85c449b007008361f10374878f2f4/pytorch_lightning/callbacks/model_checkpoint.py#L690-L691

if nothing is logged, or someone forgets to log the only metric being used for checkpointing, then it doesn't throw an error? any reason why is it done like this? cc @carmocca @awaelchli

**Comment by awaelchli** [comment]
> if nothing is logged, or someone forgets to log the only metric being used for checkpointing, then it doesn't throw an error? any reason why is it done like this?

No monitor provided will still do checkpointing, just always the last checkpoint. Without the monitor, it won't be able to track which one is the best monitor. But checkpointing is on by default.

**Comment by rohitgr7** [comment]
> > if nothing is logged, or someone forgets to log the only metric being used for checkpointing, then it doesn't throw an error? any reason why is it done like this?
> 
> No monitor provided will still do checkpointing, just always the last checkpoint. Without the monitor, it won't be able to track which one is the best monitor. But checkpointing is on by default. 

Monitor is not None here. The case is monitor is there but it's not logged by the user. Ideally it should throw an error but since nothing is logged it keeps on creating the checkpoints without and error.

**Comment by awaelchli** [comment]
It currently shows a warning. 
Probably could now also be an error since we have now two independent modes for the checkpoint callback, either training or validation. Not sure, but one has to carefully check the cases.

**Comment by Borda** [comment]
@rohitgr7 interested in sending a PR? :rabbit:

**Comment by rohitgr7** [comment]
yep, on weekend.

**Comment by tchaton** [comment]
Dear @rohitgr7,

This should be resolved with the new LoggerConnector Refactor. Might be worth to wait until then.

Best,
T.C

**Comment by carmocca** [comment]
Updated the original reproduction code to actually assert something.

This will not be resolved with #7631 as it is more of a responsibility of the loops and state tracking. Could be resolved with something like the following after the loop refactor (cc @awaelchli):

```python
    def current_loop(self) -> Loop:
        if self.state.fn == TrainerFn.FITTING:
            # train_loop or val_loop 
            return self.fit_loop.current_loop
        elif self.state.fn == TrainerFn.VALIDATING:
            return self.val_loop
        elif self.state.fn == TrainerFn.TESTING:
            return self.test_loop
        ...

    @property
    def result_collection(self) -> ResultCollection:
        return self.current_loop.collection
```

So result collection is associated with the active loop.

What we have in #7631 is:

```python
    @property
    def result_collection(self) -> Optional[ResultCollection]:
        if self.training:
            return self.train_loop.train_results
        elif self.validating or self.sanity_checking:
            return self.evaluation_loop.validation_results
        elif self.testing:
            return self.evaluation_loop.test_results
```

Which does not differentiate between the trainer function called.
